### PR TITLE
formula: enable unity builds in `std_meson_args`

### PR DESCRIPTION
**Pull request by carlocab**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

Enabling unity builds[^1][^2] can help cut down build times and can
produce faster code.

The downsides of enabling this (e.g. slower incremental rebuilds, larger
memory requirements) aren't as significant for us, so let's enable it by
default.

This can break compilation of some software, so let's add an opt-out.

[^1]: https://en.wikipedia.org/wiki/Unity_build
[^2]: https://mesonbuild.com/Unity-builds.html

### CODE REVIEW

**Comment by RandomDSdevel** [comment]
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Per the relevant section of and text in the Wikipedia article linked in this PR discussed thread's OP, enabling unity builds has potential detrimental effects on program correctness or can otherwise cause miscompilation.  If Homebrew decides to allow unity builds, it should be on a formula-by-formula basis until it can be proven that the potential compilation issues that can manifest with unity builds aren't a problem in most cases, at which point flipping the default _might_ be considered, but only if applying it wouldn't be something that formula authors should think about carefully when writing new formulas to contribute to Homebrew/homebrew-core or their own, third-party taps.  Since there's a non-trivial chance that it would be, considering making unity builds the default behavior, even for a single build system, should, in all likelihood, be met with great skepticism.  
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Also, despite the speed-ups they're reported to bring in spite of this, unity builds defeat build parallelism beyond the number of partial unity build files allocated.  (Meson's documentation, also linked to in the OP, states that the build system defaults to using 4sub-unity build jobs.)  This behavior may prove at least partially unintuitive to users who expect higher build parallelism, either generally or for specific  compilation workloads.  Both end users and CI provide machines woth multi-core CPUs; let's not under-utilize/waste them.  
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Letting formula authors request unity builds for their formula(s) could still be made easier by providing a new argument to '`std_meson_args`.'  

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(One further aside:  The kind of developers I've heard advocate for unity builds most often tend to be from a certain class of game developer.  Most formulas aren't for game software.)

**Comment by carlocab** [comment]
> Since there's a non-trivial chance that it would be

What makes you say that the chance is non-trivial? Neither of the sources I cite suggest it is.

> This behavior may prove at least partially unintuitive to users who expect higher build parallelism, either generally or for specific compilation workloads. Both end users and CI provide machines woth multi-core CPUs; let's not under-utilize/waste them.

The overwhelming majority of Homebrew users don't build formulae from source, so I don't think this is really a significant consideration here.

**Comment by eli-schwartz** [comment]
Many people have argued for and against unity builds. One argument against them is https://engineering.leewinder.co.uk/2009/12/15/the-evils-of-unity-builds/

At the end of the day, whether unity builds works is dependent on the project itself. If it does work, then there's not much cost to using that support. It will correctly compile, and due to the default chunking, you can still occupy most cores when doing a full build, since there isn't just *one* unity amalgam used.

I would be very, very wary of making this the default, because I suspect nowhere near the majority of projects compile successfully without leakage between TU-scoped code and local variables. Although this would probably just result in very visible compile errors. :sunglasses:

**Comment by RandomDSdevel** [comment]
@carlocab:  

> > Since there's a non-trivial chance that it would be
> 
> What makes you say that the chance is non-trivial? Neither of the sources I cite suggest it is.

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;From your first reference, [Wikipedia's article on unity builds](https://en.wikipedia.org/wiki/Unity_build), namely its section titled '[Effects](https://en.wikipedia.org/wiki/Unity_build#Effects):'  

> Unity builds have also potentially dangerous effects on the [semantics](https://en.wikipedia.org/wiki/Semantics_(computer_science)) of programs. Some valid C++ constructs that rely on internal linkage may fail under a unity build, for instance clashes of static symbols and symbols defined in anonymous namespaces with the same identifier in different files. If different C++ files define different functions with the same name, the compiler may unexpectedly resolve the [overloading](https://en.wikipedia.org/wiki/Function_overloading) by selecting the wrong function, in a way that was not possible when designing the software with the files as different translation units. Another adverse effect is the possible leakage of [macro](https://en.wikipedia.org/wiki/Macro_(computer_science)) definitions across different source files.[[2]](https://en.wikipedia.org/wiki/Unity_build#cite_note-2)

---

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Also, quoting @eli-schwartz since he explained the concerns I share well:  

> At the end of the day, whether unity builds works is dependent on the project itself. 

> I would be very, very wary of making this the default, because I suspect nowhere near the majority of projects compile successfully without leakage between TU-scoped code and local variables. Although this would probably just result in very visible compile errors. 😎

I could see a program slipping past this sanity check and appearing to compile just fine but actually mis-compiling not being any fun trying to debug.  It's a ticking time bomb, ready to blow up in somebody's face.  Don't take the risk; it's too much of one.  

---

> The overwhelming majority of Homebrew users don't build formulae from source, so I don't think this is really a significant consideration here.  

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;This change would also affect builds in bottle CI, would it not?  The number of end users building from source is thus irrelevant.

**Comment by MikeMcQuaid** [comment]
> The overwhelming majority of Homebrew users don't build formulae from source, so I don't think this is really a significant consideration here.

To go further: we also do not support any errors that occur when building from source.

> This change would also affect builds in bottle CI, would it not?

And it can be reverted if that's a problem.

> The number of end users building from source is thus irrelevant.

@RandomDSdevel that you personally run/ran a fork of Homebrew to build from source seems relevant to note here.

---

@carlocab I defer to whatever you think best given the above, here. It may be safer to start being opt-in or having this controlled by an environment variable that we can play around with in test-bot and/or homebrew/core but I don't have strong feelings either way. Giving this a ✅ but will leave it up to you to decide when/if to merge (but hold off until next tag regardless, thanks).

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by MikeMcQuaid** [comment]
@carlocab thoughts on merging or closing this?

**Comment by MikeMcQuaid** [comment]
Closing this out for now. @carlocab feel free to reopen whenever you're ready to merge.

**Pull request closed without merging**
